Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp

```diff
@@ -0,0 +1,168 @@
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebGPUSurfaceCocoa.h"
+
+#include "PlatformCALayer.h"
+#include "pal/graphics/WebGPU/WebGPUCanvasConfiguration.h"
+#include "pal/graphics/WebGPU/WebGPUDevice.h"
+#include "pal/graphics/WebGPU/WebGPUTextureDescriptor.h"
+
+namespace WebCore {
+
+static bool validateSurfaceCreation(IntSize, const PAL::WebGPU::CanvasConfiguration& configuration)
+{
+    // FIXME: Support RGBA8 or RGBA16Float if possible.
+    if (configuration.format != PAL::WebGPU::TextureFormat::Bgra8unorm) {
+        // FIXME: this generates a device validation error.
+        return false;
+    }
+
+    // FIXME: Support view formats
+    if (!configuration.viewFormats.isEmpty())
+        return false;
+
+    return true;
+}
+
+RefPtr<WebGPUSurfaceCocoa> WebGPUSurfaceCocoa::create(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+{
+    if (!validateSurfaceCreation(size, configuration))
+        return nullptr;
+
+    return adoptRef(*new WebGPUSurfaceCocoa(WTFMove(size), WTFMove(configuration)));
+}
+
+WebGPUSurfaceCocoa::WebGPUSurfaceCocoa(IntSize size, PAL::WebGPU::CanvasConfiguration configuration)
+    : m_size(WTFMove(size))
+    , m_configuration(WTFMove(configuration))
+{
+    ASSERT(validateSurfaceCreation(size, configuration));
+}
+
+void WebGPUSurfaceCocoa::prepareToDelegateDisplay(PlatformCALayer&)
+{
+}
+
+void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)
+{
+    if (m_frontBuffer)
+        layer.setContents(m_frontBuffer->surface());
+}
+
+GraphicsLayer::CompositingCoordinatesOrientation WebGPUSurfaceCocoa::orientation() const
+{
+    return GraphicsLayer::CompositingCoordinatesOrientation::TopDown;
+}
+
+void WebGPUSurfaceCocoa::present()
+{
+    std::swap(m_frontBuffer, m_backBuffer);
+    m_backBufferTexture = nullptr;
+    std::swap(m_backBuffer, m_secondaryBackBuffer);
```

Comment:
I don't understand how this causes the rendered results to be presented to the screen. Is that really what this function is trying to do?

Reply:
I guess this is setting up WebGPUSurfaceCocoa::display() to perform the present next time it's called. But surely present() needs to do some invalidation logic to make sure it gets called soon?

Reply:
How do we know the GPU is done rendering the frame at this point? I don't see any back pressure from the GPUP to the Web process.

Reply:
I'll take a look at this

Reply:
`void WebGPUSurfaceCocoa::display(PlatformCALayer& layer)` causes the contents to appear on the screen.

which on macOS is called from `-[WebSimpleLayer display]` and then `GraphicsLayerCA::platformCALayerLayerDisplay`

Reply:
I think the invalidation logic is handled by `GPUCanvasContext::m_compositingResultsNeedsUpdating`